src/chapter22/exercise01.lean

```diff
@@ -35,76 +41,424 @@ begin
+-- Sii is not bounded above.
+lemma part_b_ub : ¬ ∃ z : ℝ, z ∈ upper_bounds Sii :=
 begin
-  sorry
+  rw not_exists,
+  intro n,
+  rw upper_bounds,
+  simp,
+  by_cases h: n ≤ 3,
+  use 4,
+  split,
+  unfold Sii,
+  simp,
+  norm_num,
+  rw abs_of_pos,
+  norm_num,
+  norm_num,
+  linarith,
+  use n+1,
+  split,
+  unfold Sii,
+  simp,
+  rw abs_of_pos,
+  rw abs_of_pos,
+  linarith,
+  linarith,
+  linarith,
+  linarith,
 end
 
 -- if you proved there was an upper bound, then change 37 to the 
 -- right answer and prove it's a least upper bound!
 -- etc etc
-lemma part_b_lub : is_lub Sii 37 :=
-begin
-  sorry
-end
 
+-- Si is bounded below.
 lemma part_b_lb : ∃ z : ℝ, z ∈ lower_bounds Sii :=
 begin
-  sorry
+  use -2,
+  rw lower_bounds,
+  simp,
+  intro t,
+  intro h,
+  unfold Sii at h,
+  simp at h,
+  by_cases h1: t < -7,
+  rw abs_of_neg at h,
+  rw abs_of_neg at h,
+  linarith,
+  linarith,
+  linarith,
+  by_cases h2: t ≥ 3,
+  linarith,
+  rw abs_of_neg at h,
+  rw abs_of_nonneg at h,
+  linarith,
+  push_neg at h1,
+  linarith,
+  linarith,
 end
 
-lemma part_b_glb : is_glb Sii 37 :=
+-- -2 is the greatest lower bound.
+lemma part_b_glb : is_glb Sii (-2) :=
 begin
-  sorry
+  rw is_glb,
+  rw is_greatest,
+  split,
+  rw lower_bounds,
+  simp,
+  intro t,
+  intro h,
+  unfold Sii at h,
+  simp at h,
+  by_cases h1: t < -7,
+  rw abs_of_neg at h,
+  rw abs_of_neg at h,
+  linarith,
+  linarith,
+  linarith,
+  by_cases h2: t ≥ 3,
+  linarith,
+  rw abs_of_neg at h,
+  rw abs_of_nonneg at h,
+  linarith,
+  push_neg at h1,
+  linarith,
+  linarith,
+  rw upper_bounds,
+  simp,
+  intro n,
+  intro h,
+  rw lower_bounds at h,
+  simp at h,
+  unfold Sii at h,
+  by_contra h1,
+  push_neg at h1,
+  by_cases p: n≤ (-1),
+  let t:= n+2,
+  have ht: -2+t = n,
+    simp [t],
+  let k:= -2 + (t/2),
+  have l1: n ≤ k,
+  apply h,
+  simp [k],
+  rw abs_of_neg,
+  rw abs_of_pos,
+  linarith,
+  linarith,
+  linarith,
+  simp [k] at l1,
+  linarith,
+  push_neg at p,
+  have l: n≤ (-1),
+  apply h,
+  simp,
+  rw abs_of_neg,
+  rw abs_of_pos,
+  linarith,
+  linarith,
+  linarith,
+  linarith,
 end
 
 def Siii : set ℝ := {x | x^3 - 3*x < 0}
 
+-- Siii is bounded above.
 lemma part_c_ub : ∃ z : ℝ, z ∈ upper_bounds Siii :=
 begin
-  sorry
+  use (real.sqrt 3),
+  rw upper_bounds,
+  simp,
+  intro n,
+  intro h,
+  unfold Siii at h,
+  simp at h,
+  by_cases p: 0 < n,
+  have l:= div_lt_div_of_lt p h,
+  rw mul_div_assoc at l,
+  rw div_self at l,
+  simp at l,
+  have k: 1+2 = 3,
+  norm_num,
+  rw ← k at l,
+  rw pow_add at l,
+  simp at l,
+  rw mul_comm at l,
+  rw mul_div_assoc at l,
+  rw div_self at l,
+  simp at l,
+  have l1: 3 = (real.sqrt 3) ^ 2,
+  norm_num,
+  rw l1 at l,
+  rw sq_lt_sq at l,
+  have p1: 0 ≤ n,
+  linarith,
+  rw ← abs_eq_self at p1,
+  rw p1 at l,
+  have p2: 0≤ real.sqrt 3,
+  exact real.sqrt_nonneg 3,
+  rw ← abs_eq_self at p2,
+  rw p2 at l,
+  linarith,
+  linarith,
+  linarith,
```

Comment:
Take a look at how easy this can be if you use the right lemmas.
```
  { -- Let's first get h into a form where we can cancel
    have l1 : n * (n ^ 2) < n * 3,
    { convert h using 1; ring, },
    -- Now let's cancel the n
    rw mul_lt_mul_left p at l1,
    -- now let's change 3 to (sqrt 3)^2
    rw (show 3 = (real.sqrt 3) ^ 2, by norm_num) at l1,
    -- now use `sq_lt_sq` like you do
    rw sq_lt_sq at l1, 
    -- now remove the abses without creating any new goals
    rw [abs_of_nonneg p.le, abs_of_nonneg (real.sqrt_nonneg 3)] at l1,
    exact l1.le, -- dot notation is quicker than linarith.
  },
```
Do you understand my proof?